refactor (core/delivery): Convert executionGroup and executionGroups to React

jrsquared opened the pull request:
Converts executionGroup and executionGroups to React. Also created a new Sticky component to replace the angular sticky-header component in the react world.

@anotherchrisberry @christopherthielen @danielpeach PTAL

anotherchrisberry commented on app/scripts/modules/core/delivery/executionGroup/ExecutionGroup.tsx:
```diff
@@ -0,0 +1,239 @@
+      pipelineConfig,
+    };
+
+    $rootScope.$on('$stateChangeSuccess', () => {
```
A couple of issues here:
1. This won't get cleaned up when the component unmounts, so every time the state change event fires, we'll run this, even if we're on a different page.
2. We're going to be doing this a lot, so should move it into a separate service for React components to abstract this functionality.

anotherchrisberry replied:
`stateService` seems like a good place to put that abstraction

jrsquared replied:
Yep, forgot to clean this up. Very good point to just move it to stateService and sub/unsub.

anotherchrisberry commented on app/scripts/modules/core/delivery/executionGroup/ExecutionGroup.tsx:
```diff
@@ -0,0 +1,239 @@
+  private isShowingDetails(): boolean {
+    return this.props.group.executions
+          .map((execution: any) => execution.id)
+          .some((id: string) => this.showDetails(id));
```
why `.map` then `.some` and not just `.some((execution: IExecution) => this.showDetails(execution.id))`?

jrsquared replied:
Another blind copy/paste...

anotherchrisberry commented on app/scripts/modules/core/delivery/executionGroup/ExecutionGroup.tsx:
```diff
@@ -0,0 +1,239 @@
+    });
+  }
+
+  private showDetails(executionId: string): boolean {
```
`showDetails` makes me think this method shows the details, not tell you whether the details are showing for the `executionId`.

(assuming this is existing code that was badly named)

jrsquared replied:
Yep, old name that I didn't think about.

anotherchrisberry commented on app/scripts/modules/core/delivery/executionGroup/ExecutionGroup.tsx:
```diff
@@ -0,0 +1,239 @@
+  private startPipeline(command: any): IPromise<void> {
+    this.setState({triggeringExecution: true});
+    return pipelineConfigService.triggerPipeline(this.props.application.name, command.pipelineName, command.trigger).then(
+      (result: any) => {
```
I bet we can make this not be `any`

anotherchrisberry commented on app/scripts/modules/core/delivery/executionGroup/ExecutionGroups.tsx:
```diff
@@ -0,0 +1,80 @@
+};
+
+interface IState {
+  groups: any;
```
imagine how cool it would be if this weren't `any`

anotherchrisberry commented on app/scripts/modules/core/delivery/executionGroup/ExecutionGroups.tsx:
```diff
@@ -0,0 +1,80 @@
+      showingDetails: this.showingDetails()
+    }
+
+    $rootScope.$on('$stateChangeSuccess', () => {
```
same as above

anotherchrisberry commented on app/scripts/modules/core/utils/stickyHeader/Sticky.tsx:
```diff
@@ -0,0 +1,169 @@
+        {this.props.children}
+      </div>
+    );
+  }
```
Much as I'd like to pretend to have reviewed this, I can barely bring myself to look at the existing sticky header code, so am assuming it's fine and you did a good job.

jrsquared replied:
Haha!

jrsquared commented:
Addressed all comments except the $rootScope one. Also should add tests (thanks @danielpeach for writing the first react test so i can reuse all of your hard work!)

christopherthielen commented on app/scripts/modules/core/state.service.ts:
```diff
@@ -1,8 +1,27 @@
 
-export let stateService: IStateService = undefined;
-export let stateParamsService: IStateParamsService = undefined;
+export class StateService {
```
Newer version of ui-router exposes `$state` API as a class named `StateService`.   I think it would be wise to choose a different name for this.  

There is also this: https://github.com/ui-router/rx which is roughly the same thing as you wrote but only works with ui-router 1.0+

jrsquared replied:
Very good points. After talking offline, we agreed to split StateService into two things, `uirouter.ts` which is effectively the ngimport of ui-router, and `state.events` to encapsulate the rx-ing of state events.

anotherchrisberry commented on app/scripts/modules/core/api/api.service.ts:
```diff
@@ -23,6 +23,10 @@ export interface IRequestBuilder {
 
+export interface IResponse {
+  ref: string;
+}
```
I don't think this is right. The `ref` is specifically the response to the `triggerPipeline` method, and probably belongs in that service.
